File: 0013/solution.ts
function romanToInt(s: string): number {
  const entries = {
    I: 1,
    V: 5,
    X: 10,
    L: 50,
    C: 100,
    D: 500,
    M: 1000,
    IV: 4,
    IX: 9,
    XL: 40,
    XC: 90,
    CD: 400,
    CM: 900,
  };
  const symbols = new Map<string, number>(Object.entries(entries));
  let res = 0;
  for (let i = 0; i < s.length; i++) {
    if (symbols.has(s.slice(i, i + 2))) {
      res += symbols.get(s.slice(i, i + 2))!;
      i++;
    } else {
      res += symbols.get(s[i])!;
    }
  }
  return res;
}

File: 0013/solution.py
class Solution:
    def romanToInt(self, s: str) -> int:
        res = 0
        i = 0
        while i < len(s):
            if s[i] == "I":
                if s[i : i + 2] == "IV":
                    res += 4
                    i += 1
                elif s[i : i + 2] == "IX":
                    res += 9
                    i += 1
                else:
                    res += 1
            elif s[i] == "V":
                res += 5
            elif s[i] == "X":
                if s[i : i + 2] == "XL":
                    res += 40
                    i += 1
                elif s[i : i + 2] == "XC":
                    res += 90
                    i += 1
                else:
                    res += 10
            elif s[i] == "L":
                res += 50
            elif s[i] == "C":
                if s[i : i + 2] == "CD":
                    res += 400
                    i += 1
                elif s[i : i + 2] == "CM":
                    res += 900
                    i += 1
                else:
                    res += 100
            elif s[i] == "D":
                res += 500
            elif s[i] == "M":
                res += 1000
            i += 1

        return res

        # time(O(n))
        # space(O(1))
